keras/backend/jax/trainer.py

```diff
@@ -901,3 +888,49 @@ def _purge_model_variables(
         if metric_variables:
             for v in self.metrics_variables:
                 v._value = None
+
+
+def _distribute_data(data):
+    distribution = distribution_lib.distribution()
+    if distribution is not None:
+
+        def distribute_single_value(d):
+            layout = distribution.get_data_layout(d.shape)
+            return jax_distribution_lib.distribute_data_input(d, layout)
+
+        return jax.tree_util.tree_map(distribute_single_value, data)
+    else:
+        return jax.tree_util.tree_map(jax.device_put, data)
+
+
+class JAXEpochIterator(EpochIterator):
+    def _get_iterator(self, return_type="np"):
+        if return_type == "np":
+            return self._prefetch_to_device(super()._get_iterator(return_type))
+        else:
+            return super()._get_iterator(return_type)
+
+    def _prefetch_to_device(self, numpy_iterator):
+        """Shard and prefetch batches on device.
+
+        Most of the implementation has been borrowed from
+        `flax.jax_utils.prefetch_to_device`
+
+        This utility takes an iterator and returns a new iterator which fills an
+        on device prefetch buffer. Eager prefetching can improve the performance
+        of training loops significantly by overlapping compute and data
+        transfer.
+        """
+        queue = collections.deque()
+
+        # If you're training on GPUs, 2 is generally the best choice because
+        # this guarantees that you can overlap a training step on GPU with a
+        # data prefetch step on CPU.
+        def enqueue(n=2):
+            for data in itertools.islice(numpy_iterator, n):
+                queue.append(_distribute_data(data))
+
+        enqueue(n=2)  # TODO: should we make `n` configurable?
```

Comment:
A good strategy may be to make this similar to "AUTO" in tf.data, where we gradually increase the value until we hit diminishing returns (or performance reduction) and then revert back to the optimal value.

Reply:
This sounds like a more user-friendly strategy.

I have some questions about it:
1. Is it acceptable to introduce some level of overhead/complexity to fine-tune `n`?
2. `n` is only configurable at each epoch, and the statistics of one-epoch time might vary with different metrics and callbacks, etc. Is it fine to use one-epoch time to optimize `n`?
3. There is a drawback to setting a large `n`: it will consume a lot of memory on the target device.

References:
- ResNet50 example from Flax uses `2`: https://github.com/google/flax/blob/85dfad242e56098849dbf05e7e4657b3a40820f9/examples/imagenet/train.py#L214
- `torchtnt.utils.data.CudaDataPrefetcher` defaults to `2`: https://pytorch.org/tnt/stable/utils/generated/torchtnt.utils.data.CudaDataPrefetcher.html